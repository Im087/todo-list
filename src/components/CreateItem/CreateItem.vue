<template>
  <v-container class="d-flex">
    <v-text-field
      placeholder="What are you gonna do?"
      v-model="toDoItem.toDoContent"
    >
    </v-text-field>
    <v-btn color="info" size="x-large" @click="addToDoItem">Create</v-btn>
  </v-container>
</template>

<script lang="ts">
import { useStore } from "@/store";
import { defineComponent, ref, type Ref } from "vue";

import type { ToDoItem } from '../../interfaces';

export default defineComponent ({
  name: "CreateItem",
  setup() {
    const store = useStore();
    const toDoItem: Ref<ToDoItem> = ref({
      itemStatus: false,
      timeStamp: '',
      toDoContent: '',
    });

    const addToDoItem = () => {
      toDoItem.value.timeStamp = new Date().toISOString();
      let item = Object.assign({}, toDoItem.value)
      toDoItem.value.toDoContent = '';
      store.dispatch('addToDoItem', item);
    };

    return {
      toDoItem,
      addToDoItem
    };
  }
});
</script>

<style lang="scss" scoped>

</style>