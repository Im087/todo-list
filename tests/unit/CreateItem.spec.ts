import { describe, test, expect, vi, beforeEach, afterEach } from "vitest";
import CreateItem from '../../src/components/CreateItem/CreateItem.vue';
import { VueWrapper, mount } from '@vue/test-utils';
import { createVuetify } from "vuetify";
import * as components from "vuetify/components";
import * as directives from "vuetify/directives";

describe('CreateItem.vue', () => {
  const vuetify = createVuetify({components, directives});
  const $store = {
    state: {
      toDoArray: [{toDoContent: '123'}]
    },
    dispatch: vi.fn(),
    commit: vi.fn()
  };

  describe('When the input updates', () => {
    const wrapper = mount(CreateItem, {
      global: {
        plugins: [vuetify]
      }
    });

    wrapper.getComponent('[data-test-id="create-item__input"]').get('input').setValue('todo');

    test('two-way binding will work', () => {
      expect(wrapper.vm.toDoItem.toDoContent).toMatch('todo')
    });

    wrapper.unmount();
  });

  describe('When clicking on CREATE button', () => {
    let wrapper: VueWrapper<any>;
    beforeEach(() => {
      wrapper = mount(CreateItem, {
        global: {
          plugins: [vuetify],
          // mocks: {
          //   $store
          // }
        }
      });

      wrapper.getComponent('[data-test-id="create-item__input"]').get('input').setValue('todo');
    });

    afterEach(() => {
      wrapper.unmount()
    })

    test('addToDoItem function is called', async () => {
      const spy = vi.spyOn(wrapper.vm, 'addToDoItem');
      await wrapper.getComponent('[data-test-id="create-item__btn"]').get('button').trigger('click');
      expect(spy).toBeCalled();
    });
    // test('vuex action is called', () => {
    //   expect($store.dispatch).toHaveBeenCalled();
    // });
    test('input will be empty', async () => {
      await wrapper.getComponent('[data-test-id="create-item__btn"]').get('button').trigger('click');
      expect(wrapper.find('input').element.value).toEqual('');
    })
  });

  describe('snapshot', () => {
    test('is OK', () => {
      const wrapper = mount(CreateItem, {
        global: {
          plugins: [vuetify]
        }
      });
      expect(wrapper.vm.$el).toMatchSnapshot();
    });
  });
});