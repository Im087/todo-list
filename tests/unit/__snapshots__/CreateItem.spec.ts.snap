// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateItem.vue > snapshot > is OK 1`] = `
<div
  class="v-container create-item d-flex flex-wrap justify-center"
  data-v-137df678=""
>
  <div
    class="v-input v-input--horizontal v-input--density-default v-text-field create-item__input mb-2"
    data-test-id="create-item__input"
    data-v-137df678=""
  >
    <!---->
    <div
      class="v-input__control"
    >
      
      <div
        class="v-field v-field--no-label v-field--variant-filled v-theme--light"
        role="textbox"
      >
        <div
          class="v-field__overlay"
        />
        <div
          class="v-field__loader"
        >
          <div
            aria-hidden="true"
            aria-valuemax="100"
            aria-valuemin="0"
            class="v-progress-linear v-theme--light"
            role="progressbar"
            style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);"
          >
            <!---->
            <div
              class="v-progress-linear__background"
              style="width: 100%;"
            />
            <transition-stub
              appear="false"
              css="true"
              name="fade-transition"
              persisted="false"
            >
              <div
                class="v-progress-linear__indeterminate"
              >
                
                <div
                  class="v-progress-linear__indeterminate long"
                />
                <div
                  class="v-progress-linear__indeterminate short"
                />
                
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div
          class="v-field__field"
          data-no-activator=""
        >
          <!---->
          <label
            class="v-label v-field-label"
            for="input-0"
          >
            <!---->
            
            <!---->
            
          </label>
          
          
          <!---->
          <input
            aria-describedby="input-0-messages"
            class="v-field__input"
            id="input-0"
            placeholder="What are you gonna do?"
            size="1"
            type="text"
          />
          <!---->
          
          
        </div>
        <!---->
        <!---->
        <div
          class="v-field__outline"
        >
          <!---->
          <!---->
        </div>
      </div>
      
    </div>
    <!---->
    <!---->
  </div>
  <button
    class="v-btn v-btn--elevated v-theme--light bg-info v-btn--density-default v-btn--size-x-large v-btn--variant-elevated ml-6"
    data-test-id="create-item__btn"
    data-v-137df678=""
    type="button"
  >
    
    <span
      class="v-btn__overlay"
    />
    <span
      class="v-btn__underlay"
    />
    
    <!---->
    <span
      class="v-btn__content"
      data-no-activator=""
    >
      
       Create 
      
    </span>
    <!---->
    <!---->
  </button>
</div>
`;
